### ISSUE TITLE
Document Origin Timestamp (how populated, why nanoseconds, uniqueness, etc.)

### ISSUE THREAD

**Issue by jpwhitemn** [edgexfoundry/go-mod-core-contracts#562]
Per the core working group discussion of 3/25/21 - the origin timestamp (on event and reading objects) needs to be more fully documented.  Specifically, the documentation needs to address:
- Why the use of the Origin timestamp in nanosecond (vs millisecond) accuracy -> ans: some devices and use cases may provide and need that degree of accuracy)
- That Origin must be provide (required) by the device services (and how that is done when a device provides or when the device does not provide it and the device service must provide it)
- That the Origin timestamp for readings coming in at the same time must be unique by at least a nanosecond and how that is accomplished.

**Comment by cloudxxx8** [comment]
I just noticed the Origin is required in V2 now, and the Swagger mentioned it's nanoseconds.
@weichou1229 will make it more accurately in Swagger.

However, I thought the third point means event, not reading.  Is it necessary unique for readings?

**Comment by tonyespy** [comment]
@cloudxxx8 It's my understanding that it's only Origin timestamp of readings that need nanosecond accuracy,  not events (which are created by the device services themselves and thus unlikely there will be nanosecond collisions).

**Comment by cloudxxx8** [comment]
@tonyespy yes, your understanding is correct, but we should not modify the origin of reading.
we had this origin unique logic for event in v1, but not in v2 currently
https://github.com/edgexfoundry/device-sdk-go/blob/93569aefbcf2dcd4d18a31fcf2c190494ae30d88/internal/common/utils.go#L267-L276

@iain-anderson do we still need this logic to make sure the origin is unique in v2?  do we need it for event or reading?

@hahattan for awareness

**Comment by tonyespy** [comment]
@cloudxxx8 is that function called for every reading? I would think it would only apply when the Origin was set by the SDK not when set by the device service's ProtocolDriver impl.

**Comment by cloudxxx8** [comment]
@tonyespy the function was called for every Events in v1, not Reading.  The Origin set by ProtoclDriver impl won't be modified.
Thus, my point is that the Origin of Readings might be the same, but we could make sure the Origin of Event unique.

**Comment by iain-anderson** [comment]
Yes, the Event origin is always set by the SDK and it is intended to be unique for that device service instance. Collisions at nanosecond accuracy are unlikely but possible especially if the source of time on the system does not itself have nanosecond resolution, so we have some logic to avoid them.
If a Reading origin is not set by the ProtocolDriver, the SDK copies the Event origin into it. Thus collisions on Reading origin are possible and in some situations certain.

**Comment by anilchoudhury** [comment]
I was looking at the Core-Data documentation it says the device service uses millisec time resolution to record the origin time of the vents. The time is represented in int64 format.
https://docs.edgexfoundry.org/2.0/walk-through/Ch-WalkthroughReading/#Origin Timestamp.

Does Ireland release support timestamp in usec/nsec resolution. Preferebly nsec resolution? or is there any plan for it?

**Comment by cloudxxx8** [comment]
@anilchoudhury we are still working on the v2 documentation.  currently, all the document pages are copied from v1, so some of them are outdated.
In V2, we are using nsec in Reading origin timestamp.
Please open issues on edgex-docs repository if you found anything wrong.  Thanks.
